ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HiveAggregatePartitionIncrementalRewritingRule.java

```diff
@@ -0,0 +1,152 @@
+ * SELECT a, b, sum(c) sumc FROM t1 GROUP BY b, a;
+ *
+ * 1. Query all rows from source tables since the last rebuild.
+ * 2. Query all rows from MV which are in any of the partitions queried in 1.
+ * 3. Take the union of rows from 1. and 2. and perform the same aggregations defined in the MV
+ *
+ * SELECT b, sum(sumc), a FROM (
+ *     SELECT b, sumc, a FROM mat1
+ *     LEFT SEMI JOIN (SELECT b, sum(c), a FROM t1 WHERE ROW__ID.writeId > 1 GROUP BY b, a) q ON (mat1.a <=> q.a)
+ *     UNION ALL
+ *     SELECT b, sum(c) sumc, a FROM t1 WHERE ROW__ID.writeId > 1 GROUP BY b, a
+ * ) sub
+ * GROUP BY a, b
+ */
+public class HiveAggregatePartitionIncrementalRewritingRule extends RelOptRule {
+  private static final Logger LOG = LoggerFactory.getLogger(HiveAggregatePartitionIncrementalRewritingRule.class);
+
+  public static final HiveAggregatePartitionIncrementalRewritingRule INSTANCE =
+          new HiveAggregatePartitionIncrementalRewritingRule();
+
+  private HiveAggregatePartitionIncrementalRewritingRule() {
+    super(operand(Aggregate.class, operand(Union.class, any())),
+            HiveRelFactories.HIVE_BUILDER, "HiveJoinPartitionIncrementalRewritingRule");
+  }
+
+  @Override
+  public void onMatch(RelOptRuleCall call) {
+    RexBuilder rexBuilder = call.builder().getRexBuilder();
+
+    final Aggregate aggregate = call.rel(0);
+    final Union union = call.rel(1);
+    final RelNode queryBranch = union.getInput(0);
+    final RelNode mvBranch = union.getInput(1);
+
+    // find Partition col indexes in mvBranch top operator row schema
+    // mvBranch can be more complex than just a TS on the MV and the partition columns indexes in the top Operator's
+    // row schema may differ from the one in the TS row schema. Example:
+    // Project($2, $0, $1)
+    //   TableScan(table=materialized_view1, schema=a, b, part_col)
+    RelMetadataQuery relMetadataQuery = RelMetadataQuery.instance();
+    int partitionColumnCount = -1;
+    List<Integer> partitionColumnIndexes = new ArrayList<>();
+    for (int i = 0; i < mvBranch.getRowType().getFieldList().size(); ++i) {
+      RelDataTypeField relDataTypeField = mvBranch.getRowType().getFieldList().get(i);
+      RexInputRef inputRef = rexBuilder.makeInputRef(relDataTypeField.getType(), i);
+
+      Set<RexNode> expressionLineage = relMetadataQuery.getExpressionLineage(mvBranch, inputRef);
+      if (expressionLineage == null || expressionLineage.size() != 1) {
+        continue;
+      }
+
+      Set<RexTableInputRef> tableInputRefs = findRexTableInputRefs(expressionLineage.iterator().next());
+      if (tableInputRefs.size() != 1) {
+        continue;
+      }
+
+      RexTableInputRef tableInputRef = tableInputRefs.iterator().next();
+      RelOptHiveTable relOptHiveTable = (RelOptHiveTable) tableInputRef.getTableRef().getTable();
+      if (!(relOptHiveTable.getHiveTableMD().isMaterializedView())) {
+        continue;
```

Comment:
I believe this should never happen? Since this would not be safe, if it does, should you log a message and bail out completely from the rule rather than `continue`?

Reply:
Added logging and return instead of continue.